pychiquito/dsl.py

```diff
@@ -2,83 +2,159 @@
+        return Fixed(self.ast.add_fixed(name), 0)
 
-    def expose(self: CircuitContext, signal: Queriable, offset: ExposeOffset):
+    def expose(self: Circuit, signal: Queriable, offset: ExposeOffset):
+        assert self.mode == CircuitMode.SETUP
         if isinstance(signal, (Forward, Shared)):
-            self.circuit.expose(signal, offset)
+            self.ast.expose(signal, offset)
         else:
             raise TypeError(f"Can only expose ForwardSignal or SharedSignal.")
 
-    # import_halo2_advice and import_halo2_fixed are ignored.
-
-    def step_type(
-        self: CircuitContext, step_type_context: StepTypeContext
-    ) -> StepTypeContext:
-        self.circuit.add_step_type(
-            step_type_context.step_type, step_type_context.step_type.name
+    def step_type(self: Circuit, step_type: StepType) -> StepType:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.add_step_type(step_type.step_type, step_type.step_type.name)
+        return step_type
+
+    def step_type_def(self: StepType) -> StepType:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.add_step_type_def()
+
+    def fixed_gen(self: Circuit, fixed_gen_def: Callable[[FixedGenContext], None]):
+        self.ast.set_fixed_gen(fixed_gen_def)
+
+    def pragma_first_step(self: Circuit, step_type: StepType) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.first_step = step_type.step_type.id
+        print(f"first step id: {step_type.step_type.id}")
+
+    def pragma_last_step(self: Circuit, step_type: StepType) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.last_step = step_type.step_type.id
+        print(f"last step id: {step_type.step_type.id}")
+
+    def pragma_num_steps(self: Circuit, num_steps: int) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.num_steps = num_steps
+
+    def pragma_disable_q_enable(self: Circuit) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.q_enable = False
+
+    def gen_witness(self: Circuit, args: Any) -> TraceWitness:
+        self.mode = CircuitMode.Trace
+        self.trace_context = TraceContext()
+        self.ast.set_trace(self.trace)
+        self.trace(args)
+        return self.trace_context.witness
+
+    def add(self: Circuit, step_type: StepType, args: Any):
+        print(self)
+        print(step_type)
+        print(args)
+        assert self.mode == CircuitMode.Trace
+        self.trace_context.add(self, step_type, args)
+
+    def print_ast(self: Circuit):
+        print("Print ASTCircuit using custom __str__ method in python:")
+        print(self.ast)
+
+    def get_ast_json(self: Circuit, print_json=False) -> str:
+        ast_json: str = json.dumps(self.ast, cls=CustomEncoder, indent=4)
+        if print_json:
+            print("Print ASTCircuit using __json__ method in python:")
+            print(ast_json)
+        return ast_json
+
+    def print_witness(self: Circuit):
```

Comment:
@qwang98 there is no one witness per circuit, one circuit can generate many witnesses

Reply:
This is an user interface, so I removed it from Circuit methods.